#167190540 Implement Follow And Unfollow User Feature

henryade opened the pull request:
#### What does this PR do?
It allows users to follow and unfollow other users, get all followers and get all users they follow.

#### Description of Task proposed in this pull request?
- create user follower model
- create follow user route
- create unfollow user route
- create get followers route
- create get followings route
- write end to end test on all route
- write model unit tests

#### How should this be manually tested (Quality Assurance)?
- clone repo and checkout to this branch
- signup two users and copy the first users token and the second user's username
### To Follow
- run `POST /api/v1/profiles/{username}/follow`add the first users token to the authorization header
### To get All followers
- run `GET /api/v1/profiles/followers` add the first users token to the authorization header
### To get All followings
- run `GET /api/v1/profiles/followings` add the first users token to the authorization header
### To Unfollow
- run `DELETE /api/v1/profiles/{username}/follow`add the first users token to the authorization header

#### What are the relevant pivotal tracker stories?
- Story ID:  [#167190540](https://www.pivotaltracker.com/story/show/167190540)

#### Any background context you want to add (Operations Impact)?
N/A

#### What I have learned working on this feature: 
- I learned how to write model tests and add custom error messages to models.
- I learned how to use sequelize magic methods. 

#### Screenshots: 
User Successfully follows another user:
<img width="1440" alt="Screenshot 2019-08-08 at 4 27 49 PM" src="https://user-images.githubusercontent.com/47576037/62716399-09452600-b9fa-11e9-871a-8831a1ce4b3e.png">


User Successfully unfollow another user:
<img width="1440" alt="Screenshot 2019-08-08 at 4 31 34 PM" src="https://user-images.githubusercontent.com/47576037/62716446-1b26c900-b9fa-11e9-998d-f6eafb500132.png">


Other User View All Followers:
<img width="1440" alt="Screenshot 2019-08-08 at 4 31 03 PM" src="https://user-images.githubusercontent.com/47576037/62716430-15c97e80-b9fa-11e9-86b9-86f0bd49e721.png">


User View All Followings:
<img width="1440" alt="Screenshot 2019-08-08 at 4 28 22 PM" src="https://user-images.githubusercontent.com/47576037/62716417-1104ca80-b9fa-11e9-937c-fd4a57885852.png">

dinobi commented on server/controllers/Followers.js:
```diff
@@ -0,0 +1,151 @@
+        params: { username }
+      } = req;
+      const user = await User.findByUsername(username);
+      if (!user) {
```
@henryade don't you think we should move these checks from lines 27-34 to another method inside this controller and use it here and also in the unfollow route handler, cos it looks like we are repeating code. In the future we could move to a helper, if we discover that someone else is trying to do this same thing.

henryade replied:
Okay. I will do this now.

dinobi commented on server/controllers/Followers.js:
```diff
@@ -0,0 +1,151 @@
+      const {
+        user: { id }
+      } = req;
+      const user = await User.findByPk(id);
```
I thought we defined a `findByID`?

henryade replied:
Yes. I will change this.

dinobi commented on server/controllers/Followers.js:
```diff
@@ -0,0 +1,151 @@
+          }
+        ]
+      });
+      const users = followers.map((x) => {
```
never use single characters for parameter names unless it's a counter, change to `x` to `follower`

henryade replied:
okay. Noted.

chialuka commented on server/database/migrations/20190805160738-create-user-follower.js:
```diff
@@ -0,0 +1,37 @@
+module.exports = {
```
To be consistent, could you please use ES6 syntax here and in the other migration files?

henryade replied:
Thank you

chialuka reviewed (dismissed):
Good job.

henryade committed:
ft-[167190540]: implement follow user feature
- create user follower model
- create follow user route
- create unfollow user route
- create get followers route
- create get followings route
- write end to end test on all route
- write model unit tests

[Delivers #167190540]
